### PR TITLE
Coerce integer-indexed maps to slices

### PR DESCRIPTION
**Pull request by brandur**
This is unusual for normal "Rack-style" form encoding, but at Stripe on
some endpoints we allow maps that are indexed with integers to be
changed to arrays in the backend. So for example this:

    myarr[0]=foo

Would be interpreted in the API as:

    ["foo"]

The purpose of this is to be able to use arbitrary indexing to update
only certain items, so this also works:

    myarr[3]=foo

Becomes:

    [nil, nil, nil "foo"]

You can find an example of this in the `items` parameter on create
subscription.

Here we augment the coercer to detect if the target type is `array` and
then try to coerce a map to a slice as long as all its indexes were
strings that looked like integers.

I've added a test suite, but have also verified that this fixes a
failing test in my stripe-go PR that's targeting stripe-mock.

r? @tmaxwell-stripe Mind taking a look? Sorry I know this messes with a PR you
have in flight, but hopefully the rebase won't be too bad. We're basically
adding a function and a couple new const/vars.

### CODE REVIEW

File: param/coercer/coercer_test.go [hunk 1]
```diff
@@ -27,10 +28,75 @@ func TestCoerceParams_IntegerCoercion(t *testing.T) {
...
+		data := map[string]interface{}{
+			"arraykey": map[string]interface{}{
+				"0": "0-index",
+				"2": "2-index",
+			},
+		}
+
+		err := CoerceParams(schema, data)
+		assert.NoError(t, err)
+		assert.Equal(t, "0-index", data["arraykey"].([]interface{})[0])
```
Comment (tmaxwell-stripe) [inline param/coercer/coercer_test.go]: Would be nice to also check `2-index`

**Comment by tmaxwell-stripe** [comment]
Other than one minor ask, it looks good to me

File: param/coercer/coercer_test.go (hunk 1 above)
Comment (brandur-stripe) [reply param/coercer/coercer_test.go]: Cool! Added that and also another test that the 1st element is a `nil` as expected.

**Comment by brandur-stripe** [comment]
Pushed change as v0.1.23.
